### DISCUSSION TITLE
Question about the derivation of projection matrix formula

### DISCUSSION THREAD

**Discussion by cmusjtuliuyuan** [InteractiveComputerGraphics/PositionBasedDynamics#189] [Q&A]
Hello, I'm a new-bee of PBD. I'm trying my best to implement PBD in CUDA. 
In PositionBasedRigidBodyDynaics.cpp, I am confused about how to derive the formula of Pr in HingeJoint. To be precise, I do not understand how the following formulas are derived.
`Eigen::Matrix<Real, 2, 4, Eigen::DontAlign> Pr = (QHatq10.transpose() * Qq00).block<2, 4>(2, 0);`
`const Eigen::Matrix<Real, 2, 3> t = -Pr * (Qq0.transpose() * Gq1);`
Is there any paper about how to derive these？

**Comment by janbender** [comment]
No, unfortunately we never published this derivation. The function constrains the two rotational degrees of freedom between the connected rigid bodies.

**Reply by cmusjtuliuyuan** [reply]
Thank you for your response. So, does the right-lower term of Jacobian t imply that the two rotational degrees of freedom between the two rigid bodies should be aligned?"

**Comment by janbender** [comment]
Yes. The relative rotation of both bodies is constrained. However, since we don't want to constrain all 3 rotational degrees of freedom, the contraint is projected to the two-dimensional space of the 2 DOFs which should be constrained.
